### ISSUE TITLE
Add Option for External Syntax Highlighter

### ISSUE THREAD

**Issue by tajmone** [jgm/pandoc#3353]
I'm thrilled by the perspectives mentioned in #3334 — skylighting is promising!

But I think it would still be nice to have an option to redirect syntax highlighting to an external tool — eg: **Pygments**, or Simon André's **Highlight**, and so on ...

Not really sure how this could be constructed to handle all cases, but I guess that it would require that the user specifies a command that pandoc should send to the shell, followed by the actual code block, and that in turn pandoc should receive the highilighted text stream back from STDIN.

Something like this:

```
pandoc --external-highlighter=highlighter_tool_name
```

For example, with **Highlight** if no input- and output-file are specified, it will uses STDIN and STDOUT by default. So, assuming that `highilight.exe` (under Windows) was on `%PATH%`, the above example should work if pandoc were to send the raw source code block to STDIN and wait for the result on STDOUT.

I assume that UTF8 should be standard format for the stream, but some extra options could allow tweaking this for special use cases.

Does it make sense? Or would it be better to have some config file to point to?

I think this would be a really nice feature, especially for HTML-based outputs. Currently, minor languages which don't have a syntax definition hardcoded in pandoc haven't much choice but to rely on in-browser JavaScript solutions (like **Highlight.js**). 

There are also cases where one might need some special syntax highlighting which is best implemented through a CLI tool like **Highlight**, which can be fed custom language definition files on the fly  — for example, in tutorials where one wants to include wrongly written code examples, for which a tweaked language syntax (like `python_err`) can be used along with the default one, to provide different coloring and highlight the mistakes, or use a modded language definition (like `python_syntax`)  for syntax rules (with all the `[`, `|` and `<` symbols that surround parameters and optional stuff).

CLI highlighter are easier to use in non-standard scenarios (no need to rebuild a library), so if pandoc could open a door to redirecting source code blocks to external tools it would be great.

**Comment by crmackay** [comment]
you could turn off highlighting in pandoc with `--no-highlight`, and use a filter to send out each code block to whatever program you want and insert the result in it's place.

**Comment by tajmone** [comment]
Good point. I'll work on it using Lua then. If I get anywhere with it I'll share it on pandoc's third-party resources page. Thanks for the tip @crmackay !

I still think it's a feature worth of inclusion in base pandoc though — something that shouldn't require external scripts, but be integrated within pandoc.

**Comment by crmackay** [comment]
you can write a filter in any language you want, it just needs to be executable, accept json from stdin and emit json from stdout.

I think many people go with python because [pandocfilters](https://github.com/jgm/pandocfilters) and [panflute](https://github.com/sergiocorreia/panflute) libraries offer good support and make it much easier to work with the AST.

I think it would be tricky to make this a builtin standard option though because there is no standardized way to pass the block of code, the language you want it parsed as, and the style you want.

**Comment by jgm** [comment]
Yes, this is definitely a job for a filter.

I don't want to add this to pandoc as an option.  For one
thing, the output of these CLI tools can be very different
from what pandoc would expect. In general you may need to
tweak things depending on the precise CLI tool you use, so
writing a custom filter is the best approach. A filter that
does this is easy, and there are examples in the panflute
repository that do very similar things.

**Comment by jgm** [comment]
By the way, skylighting can also

> be fed custom language definition files on the fly

and we'll be adding that ability to pandoc soon.

**Comment by tajmone** [comment]
@crmackay

> I think it would be tricky to make this a builtin standard option though because there is no standardized way to pass the block of code, the language you want it parsed as, and the style you want.

Mhhh ... I didn't consider that the language variable had to be passed on too. True, a filter can mediate the whole issue far better and target specific tools.

Hi @jgm ! Yes, I've read that in issue #3334, and it sounds really promising.

I've tried to peek into skylighting repo, but today I had problems accessing GitHub (servers were unreachable from Italy, on and off, all afternoon). But I've managed to view the `README`.

From what I understand, it will keep using Kate's language definitions, so it isn't going to break backward compatiblity, but it will handle them better. This is definitely good news for «minor» languages which won't make it into the supported languages list — and for which there might already be a Kate syntax file.

I've finally found the time to install Ubuntu on a second drive partition, so now I might also dig into Haskell with the freedom of *nix systems. And I might soon be digging deeper into pandoc sources ... just need to get around the great scare of Haskell that has been keeping me off (beside time issues and Windows making use of Haskell more complex than needed).

**Comment by tajmone** [comment]
I managed to achieve the goal by using [CDSoft's PP](https://github.com/CDSoft/pp) (Generic preprocessor with pandoc in mind). Since pp macros can accept parameters, specifying both the external file and language syntax was a piece of cake. 

The markdown source file is preprocessed by PP, which expands all macros: the Highlight macros inject into the document the raw html output from Highlight. The markdown output of PP is piped to pandoc and converted to html (the actual macro is for html only).

Since pandoc doesnt' highlight raw html code blocks, there is no need to disable syntax highlighting, and it's possible to have fenced code highlighted by pandoc, and external source files highlighted via Highlight: beside the fact that the two rely on different CSS styling, there are no conflicts whatsoever.

I've published the macros within a pp-macros library project (MIT License):

https://github.com/tajmone/pandoc-goodies/tree/master/pp#highlight
